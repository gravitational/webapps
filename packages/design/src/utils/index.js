/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import ReactDOM from 'react-dom';
import getOffset from 'dom-helpers/query/offset';
import getPosition from 'dom-helpers/query/position';
import getScrollTop from 'dom-helpers/query/scrollTop';

export function ownerDocument(node) {
  return (node && node.ownerDocument) || document;
}

export function ownerWindow(node, fallback = window) {
  const doc = ownerDocument(node);
  return doc.defaultView || doc.parentView || fallback;
}

export function getContainer(container, defaultContainer) {
  container = typeof container === 'function' ? container() : container;
  return ReactDOM.findDOMNode(container) || defaultContainer;
}

function getContainerDimensions(containerNode) {
  let width, height, scroll;

  if (containerNode.tagName === 'BODY') {
    width = window.innerWidth;
    height = window.innerHeight;

    scroll =
      getScrollTop(ownerDocument(containerNode).documentElement) ||
      getScrollTop(containerNode);
  } else {
    ({ width, height } = getOffset(containerNode));
    scroll = getScrollTop(containerNode);
  }

  return { width, height, scroll };
}

function getTopDelta(top, overlayHeight, container, padding) {
  const containerDimensions = getContainerDimensions(container);
  const containerScroll = containerDimensions.scroll;
  const containerHeight = containerDimensions.height;

  const topEdgeOffset = top - padding - containerScroll;
  const bottomEdgeOffset = top + padding - containerScroll + overlayHeight;

  if (topEdgeOffset < 0) {
    return -topEdgeOffset;
  } else if (bottomEdgeOffset > containerHeight) {
    return containerHeight - bottomEdgeOffset;
  } else {
    return 0;
  }
}

function getLeftDelta(left, overlayWidth, container, padding) {
  const containerDimensions = getContainerDimensions(container);
  const containerWidth = containerDimensions.width;

  const leftEdgeOffset = left - padding;
  const rightEdgeOffset = left + padding + overlayWidth;

  if (leftEdgeOffset < 0) {
    return -leftEdgeOffset;
  } else if (rightEdgeOffset > containerWidth) {
    return containerWidth - rightEdgeOffset;
  }

  return 0;
}

export function calculatePosition(
  placement,
  overlayNode,
  target,
  container,
  padding
) {
  const childOffset =
    container.tagName === 'BODY'
      ? getOffset(target)
      : getPosition(target, container);

  const { height: overlayHeight, width: overlayWidth } = getOffset(overlayNode);

  let positionLeft, positionTop, arrowOffsetLeft, arrowOffsetTop;

  if (placement === 'left' || placement === 'right') {
    positionTop = childOffset.top + (childOffset.height - overlayHeight) / 2;

    if (placement === 'left') {
      positionLeft = childOffset.left - overlayWidth;
    } else {
      positionLeft = childOffset.left + childOffset.width;
    }

    const topDelta = getTopDelta(
      positionTop,
      overlayHeight,
      container,
      padding
    );

    positionTop += topDelta;
    arrowOffsetTop = 50 * (1 - (2 * topDelta) / overlayHeight) + '%';
    arrowOffsetLeft = void 0;
  } else if (placement === 'top' || placement === 'bottom') {
    positionLeft = childOffset.left + (childOffset.width - overlayWidth) / 2;

    if (placement === 'top') {
      positionTop = childOffset.top - overlayHeight;
    } else {
      positionTop = childOffset.top + childOffset.height;
    }

    const leftDelta = getLeftDelta(
      positionLeft,
      overlayWidth,
      container,
      padding
    );

    positionLeft += leftDelta;
    arrowOffsetLeft = 50 * (1 - (2 * leftDelta) / overlayWidth) + '%';
    arrowOffsetTop = void 0;
  } else {
    throw new Error(
      `calcOverlayPosition(): No such placement of "${placement}" found.`
    );
  }

  return { positionLeft, positionTop, arrowOffsetLeft, arrowOffsetTop };
}
