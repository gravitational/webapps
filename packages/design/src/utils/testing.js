import React from 'react';
import {
  render as testingRender,
  screen,
  fireEvent,
} from '@testing-library/react';
import ThemeProvider from 'design/ThemeProvider';
import theme from 'design/theme';
import 'jest-styled-components';

export function render(component) {
  return testingRender(
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  );
}

export { screen, fireEvent, theme };
