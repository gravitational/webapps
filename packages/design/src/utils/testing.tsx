/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  render as testingRender,
  act,
  fireEvent,
  waitForElement,
} from '@testing-library/react';
import { screen, wait, prettyDOM, getByTestId } from '@testing-library/dom';
import { MemoryRouter as Router } from 'react-router-dom';
import ThemeProvider from 'design/ThemeProvider';
import theme from 'design/theme';
import '@testing-library/jest-dom';
import 'jest-styled-components';

function Providers({ children }: { children: React.ReactElement }) {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
}

function render(ui: React.ReactElement<any>, options?: RenderOptions) {
  return testingRender(ui, { wrapper: Providers, ...options });
}

screen.debug = () => {
  window.console.log(prettyDOM());
};

type RenderOptions = {
  wrapper: React.FC;
  container: HTMLElement;
};

export {
  act,
  screen,
  wait,
  fireEvent,
  theme,
  render,
  prettyDOM,
  waitForElement,
  getByTestId,
  Router,
};
