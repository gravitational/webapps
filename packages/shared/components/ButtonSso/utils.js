/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import * as Icons from 'design/Icon';
import { AuthProviderTypeEnum } from '../../services/enums';

export const TypeEnum = {
  MICROSOFT: 'microsoft',
  GITHUB: 'github',
  BITBUCKET: 'bitbucket',
  GOOGLE: 'google',
};

export function pickSsoIcon(type) {
  switch (type) {
    case TypeEnum.MICROSOFT:
      return { color: '#2672ec', Icon: Icons.Windows, type };
    case TypeEnum.GITHUB:
      return { color: '#444444', Icon: Icons.Github, type };
    case TypeEnum.BITBUCKET:
      return { color: '#205081', Icon: Icons.BitBucket, type };
    case TypeEnum.GOOGLE:
      return { color: '#dd4b39', Icon: Icons.Google, type };
    default:
      // provide default icon for unknown social providers
      return { color: '#f7931e', Icon: Icons.OpenID };
  }
}

export function guessProviderType(name, ssoType) {
  name = name.toLowerCase();

  if (name.indexOf(TypeEnum.MICROSOFT) !== -1) {
    return TypeEnum.MICROSOFT;
  }

  if (name.indexOf(TypeEnum.BITBUCKET) !== -1) {
    return TypeEnum.BITBUCKET;
  }

  if (name.indexOf(TypeEnum.GOOGLE) !== -1) {
    return TypeEnum.GOOGLE;
  }

  if (
    name.indexOf(TypeEnum.GITHUB) !== -1 ||
    ssoType === AuthProviderTypeEnum.GITHUB
  ) {
    return TypeEnum.GITHUB;
  }

  if (ssoType === AuthProviderTypeEnum.OIDC) {
    return 'openid';
  }

  return '--unknown';
}
