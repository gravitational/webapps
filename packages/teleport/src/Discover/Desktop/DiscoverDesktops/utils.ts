import { Path } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/Path';
import { MonotoneX } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/MonotoneX';

const STROKE_WIDTH = 4;

export interface Line {
  width: number;
  height: number;
  path: string;
}

export function createLine(
  desktopServiceElement: HTMLDivElement,
  desktopElement: HTMLDivElement,
  containerElement: HTMLDivElement
): Line {
  if (!desktopElement || !desktopServiceElement || !containerElement) {
    return null;
  }

  const desktopServiceRect = desktopServiceElement.getBoundingClientRect();
  const desktopRect = desktopElement.getBoundingClientRect();
  const containerRect = containerElement.getBoundingClientRect();

  const distance = desktopRect.left - desktopServiceRect.right;

  const path = new Path();
  const line = new MonotoneX(path);

  line.lineStart();

  const desktopLinePosition =
    desktopRect.top - containerRect.top + desktopRect.height / 2 - 1;
  const desktopServiceLinePosition =
    desktopServiceRect.top - containerRect.top + desktopServiceRect.height / 2;

  line.point(0, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(40, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(distance - 10, desktopLinePosition + STROKE_WIDTH / 2);
  line.point(distance, desktopLinePosition + STROKE_WIDTH / 2);

  line.lineEnd();

  return {
    width: distance,
    height: containerRect.height,
    path: path.path,
  };
}
