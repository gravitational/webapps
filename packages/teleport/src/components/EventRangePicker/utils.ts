import moment from 'moment';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: moment(new Date())
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: '7 days',
      from: moment()
        .subtract(6, 'day')
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
