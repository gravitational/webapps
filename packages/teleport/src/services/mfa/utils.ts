/**
 * Copyright 2021 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Auth2faType, PreferredMfaType } from 'shared/services';

export type MfaOption = {
  value: Auth2faType;
  label: string;
};

export function getMfaOptions(
  mfa: Auth2faType,
  preferredMfa: PreferredMfaType,
  requireMfa = false
) {
  const mfaOptions: MfaOption[] = [];

  if (mfa === 'off' || !mfa) {
    return mfaOptions;
  }

  const mfaEnabled = mfa === 'on' || mfa === 'optional';
  const preferMfaWebauthn = mfaEnabled && preferredMfa === 'webauthn';
  const preferMfaU2f = mfaEnabled && preferredMfa === 'u2f';

  if (mfa === 'webauthn' || preferMfaWebauthn) {
    mfaOptions.push({ value: 'webauthn', label: 'Hardware Key' });
  }

  if (mfa === 'u2f' || preferMfaU2f) {
    mfaOptions.push({ value: 'u2f', label: 'Hardware Key' });
  }

  if (mfa === 'otp' || mfaEnabled) {
    mfaOptions.push({ value: 'otp', label: 'Authenticator App' });
  }

  if (!requireMfa && mfa === 'optional') {
    mfaOptions.push({ value: 'optional', label: 'None' });
  }

  return mfaOptions;
}
