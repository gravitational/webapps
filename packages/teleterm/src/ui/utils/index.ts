export * from './uid';
export * from './retryWithRelogin';
export * from './getUserWithClusterName';
