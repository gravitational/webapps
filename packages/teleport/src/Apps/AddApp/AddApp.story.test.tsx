/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  Created,
  Loaded,
  Failed,
  ManuallySSO,
  ManuallyLocal,
} from './AddApp.story';
import { render, screen } from 'design/utils/testing';

test('render automatic tab', async () => {
  render(<Loaded />);
  expect(screen.getByTestId('Modal')).toMatchSnapshot();
});

test('render failed automatic tab', async () => {
  render(<Failed />);
  expect(screen.getByTestId('Modal')).toMatchSnapshot();
});

test('render automatic tab on created state', async () => {
  render(<Created />);
  expect(screen.getByTestId('Modal')).toMatchSnapshot();
});

test('render manual tab with local user', async () => {
  render(<ManuallyLocal />);
  expect(screen.getByTestId('Modal')).toMatchSnapshot();
});

test('render manual tab with sso user', async () => {
  render(<ManuallySSO />);
  expect(screen.getByTestId('Modal')).toMatchSnapshot();
});
