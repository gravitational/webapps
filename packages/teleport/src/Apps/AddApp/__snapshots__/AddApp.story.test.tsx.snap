// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`created 1`] = `
.c9 {
  box-sizing: border-box;
  margin-bottom: 24px;
  margin-right: 16px;
  width: 320px;
}

.c12 {
  box-sizing: border-box;
  margin-bottom: 24px;
  width: 100%;
}

.c17 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
}

.c17:active {
  opacity: 0.56;
}

.c17:hover,
.c17:focus {
  background: #651FFF;
}

.c17:active {
  background: #354AA4;
}

.c17:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c19 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  margin-right: 16px;
}

.c19:active {
  opacity: 0.56;
}

.c19:hover,
.c19:focus {
  background: #651FFF;
}

.c19:active {
  background: #354AA4;
}

.c19:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c20 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #222C59;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
}

.c20:active {
  opacity: 0.56;
}

.c20:hover,
.c20:focus {
  background: #2C3A73;
}

.c20:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c11 {
  appearance: none;
  border: none;
  border-radius: 4px;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  color: #324148;
  background-color: #FFFFFF;
}

.c11::-ms-clear {
  display: none;
}

.c11::placeholder {
  opacity: 0.4;
}

.c11:read-only {
  cursor: not-allowed;
}

.c10 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 4px;
}

.c6 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  text-transform: uppercase;
  margin: 0px;
  margin-right: auto;
  color: rgba(255,255,255,0.87);
}

.c13 {
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0px;
  margin-bottom: 16px;
}

.c14 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 600;
  margin: 0px;
}

.c16 {
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0px;
  margin-right: 16px;
}

.c4 {
  box-sizing: border-box;
  flex: 1;
  display: flex;
  flex-direction: column;
}

.c5 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.c8 {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  flex-direction: row;
}

.c15 {
  box-sizing: border-box;
  margin-bottom: 8px;
  padding: 8px;
  background-color: #010B1C;
  border-radius: 4px;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.c1 {
  z-index: -1;
  position: fixed;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
  background-color: rgba(0,0,0,0.5);
  opacity: 1;
  touch-action: none;
}

.c0 {
  position: fixed;
  z-index: 1200;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
}

.c2 {
  height: 100%;
  outline: none;
  color: black;
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: 1;
  will-change: opacity;
  transition: opacity 225ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c3 {
  padding: 32px;
  padding-top: 24px;
  background: #1C254D;
  color: rgba(255,255,255,0.87);
  border-radius: 8px;
  box-shadow: 0 8px 32px rgba(0,0,0,0.24);
  display: flex;
  flex-direction: column;
  position: relative;
  overflow-y: auto;
  max-height: calc(100% - 96px);
  max-width: 600px;
  width: 100%;
  min-height: 330px;
}

.c7 {
  box-sizing: border-box;
  margin-bottom: 32px;
  min-height: 254px;
  flex: 0 0 auto;
  display: flex;
  flex-direction: column;
}

.c18 {
  box-sizing: border-box;
}

<div
  class="c0"
  data-testid="Modal"
>
  <div
    aria-hidden="true"
    class="c1"
    data-testid="backdrop"
  />
  <div
    class="c2"
  >
    <div
      class="c3"
      data-testid="dialogbox"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <div
            class="c6"
            color="text.primary"
          >
            Add Application
          </div>
        </div>
        <div
          class="c7"
        >
          <div
            class="c8"
          >
            <div
              class="c9"
              width="320px"
            >
              <label
                class="c10"
                font-size="0"
              >
                App Name
              </label>
              <input
                autocomplete="off"
                class="c11"
                color="text.onLight"
                placeholder="jenkins"
                type="text"
                value=""
              />
            </div>
            <div
              class="c12"
              width="100%"
            >
              <label
                class="c10"
                font-size="0"
              >
                INTERNAL APPLICATION URL
              </label>
              <input
                autocomplete="off"
                class="c11"
                color="text.onLight"
                placeholder="https://localhost:4000"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="c13"
          >
            Use the script below to add an application to your cluster.
             
            <br />
            The script will be valid for
            <span
              class="c14"
            >
               1 hour
              .
            </span>
          </div>
          <div
            class="c15"
          >
            <div
              class="c16"
              style="word-break: break-all; font-size: 12px; font-family: \\"Droid Sans Mono\\", \\"monospace\\", monospace, \\"Droid Sans Fallback\\";"
            >
              sudo bash -c "$(curl -fsSL 'http://localhost/scripts/86/install-app.sh?name=test&uri=http://myapp/')"
            </div>
            <button
              class="c17"
              kind="primary"
              style="padding: 4px 8px; min-height: 10px; font-size: 10px;"
            >
              Copy
            </button>
          </div>
        </div>
        <div
          class="c18"
        >
          <button
            class="c19"
            kind="primary"
          >
            Regenerate
          </button>
          <button
            class="c20"
            kind="secondary"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`loaded 1`] = `
.c9 {
  box-sizing: border-box;
  margin-bottom: 24px;
  margin-right: 16px;
  width: 320px;
}

.c12 {
  box-sizing: border-box;
  margin-bottom: 24px;
  width: 100%;
}

.c15 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  margin-right: 16px;
}

.c15:active {
  opacity: 0.56;
}

.c15:hover,
.c15:focus {
  background: #651FFF;
}

.c15:active {
  background: #354AA4;
}

.c15:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c16 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #222C59;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
}

.c16:active {
  opacity: 0.56;
}

.c16:hover,
.c16:focus {
  background: #2C3A73;
}

.c16:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c11 {
  appearance: none;
  border: none;
  border-radius: 4px;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  color: #324148;
  background-color: #FFFFFF;
}

.c11::-ms-clear {
  display: none;
}

.c11::placeholder {
  opacity: 0.4;
}

.c11:read-only {
  cursor: not-allowed;
}

.c10 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 4px;
}

.c6 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  text-transform: uppercase;
  margin: 0px;
  margin-right: auto;
  color: rgba(255,255,255,0.87);
}

.c13 {
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0px;
  margin-bottom: 16px;
}

.c4 {
  box-sizing: border-box;
  flex: 1;
  display: flex;
  flex-direction: column;
}

.c5 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.c8 {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  flex-direction: row;
}

.c1 {
  z-index: -1;
  position: fixed;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
  background-color: rgba(0,0,0,0.5);
  opacity: 1;
  touch-action: none;
}

.c0 {
  position: fixed;
  z-index: 1200;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
}

.c2 {
  height: 100%;
  outline: none;
  color: black;
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: 1;
  will-change: opacity;
  transition: opacity 225ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c3 {
  padding: 32px;
  padding-top: 24px;
  background: #1C254D;
  color: rgba(255,255,255,0.87);
  border-radius: 8px;
  box-shadow: 0 8px 32px rgba(0,0,0,0.24);
  display: flex;
  flex-direction: column;
  position: relative;
  overflow-y: auto;
  max-height: calc(100% - 96px);
  max-width: 600px;
  width: 100%;
  min-height: 330px;
}

.c7 {
  box-sizing: border-box;
  margin-bottom: 32px;
  min-height: 254px;
  flex: 0 0 auto;
  display: flex;
  flex-direction: column;
}

.c14 {
  box-sizing: border-box;
}

<div
  class="c0"
  data-testid="Modal"
>
  <div
    aria-hidden="true"
    class="c1"
    data-testid="backdrop"
  />
  <div
    class="c2"
  >
    <div
      class="c3"
      data-testid="dialogbox"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <div
            class="c6"
            color="text.primary"
          >
            Add Application
          </div>
        </div>
        <div
          class="c7"
        >
          <div
            class="c8"
          >
            <div
              class="c9"
              width="320px"
            >
              <label
                class="c10"
                font-size="0"
              >
                App Name
              </label>
              <input
                autocomplete="off"
                class="c11"
                color="text.onLight"
                placeholder="jenkins"
                type="text"
                value=""
              />
            </div>
            <div
              class="c12"
              width="100%"
            >
              <label
                class="c10"
                font-size="0"
              >
                INTERNAL APPLICATION URL
              </label>
              <input
                autocomplete="off"
                class="c11"
                color="text.onLight"
                placeholder="https://localhost:4000"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="c13"
          >
            Provide the name and URL of your application to generate a script that will automatically install and configure the App service on the server that can access your application.
          </div>
          
        </div>
        <div
          class="c14"
        >
          <button
            class="c15"
            kind="primary"
          >
            Generate Script
          </button>
          
          <button
            class="c16"
            kind="secondary"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
